test/unit/gateways/versa_pay_test.rb

```diff
@@ -0,0 +1,812 @@
+require 'test_helper'
+
+class VersaPayTest < Test::Unit::TestCase
+  include CommStub
+
+  def setup
+    @gateway = VersaPayGateway.new(fixtures(:versa_pay))
+    @credit_card = credit_card
+    @amount = 100
+    @options = {
+      email: 'test@gmail.com',
+      billing_address: address.merge(name: 'Cure Tester'),
+      ip_address: '127.0.0.1'
+    }
+  end
+
+  def test_required_client_id_and_client_secret
+    error = assert_raises ArgumentError do
+      VersaPayGateway.new
+    end
+
+    assert_equal 'Missing required parameter: api_token', error.message
+  end
+
+  def test_supported_card_types
+    assert_equal VersaPayGateway.supported_cardtypes, %i[visa master american_express discover]
+  end
+
+  def test_supported_countries
+    assert_equal VersaPayGateway.supported_countries, ['US']
+  end
+
+  def test_request_headers_building
+    gateway = VersaPayGateway.new(api_token: 'abc123', api_key: 'def456')
+    headers = gateway.send :request_headers
+
+    assert_equal 'application/json', headers['Content-Type']
+    assert_equal 'Basic YWJjMTIzOmRlZjQ1Ng==', headers['Authorization']
+  end
+
+  def test_build_order_request_url
+    action = :auth
+    assert_equal @gateway.send(:url, action), "#{@gateway.test_url}/api/gateway/v1/orders/auth"
+  end
+
+  def test_error_code_from_errors
+    # a HTTP 412 response structure
+    error = @gateway.send(:error_code_from, { 'success' => false, 'errors' => ['fund_address_unspecified'], 'response_code' => 999 })
+    assert_equal error, 'response_code: 999'
+  end
+
+  def test_error_code_from_gateway_error_code
+    error = @gateway.send(:error_code_from, declined_errors)
+    assert_equal error, 'gateway_error_code: 567.005 | response_code: 999'
+  end
+
+  def test_message_from_successful_purchase
+    message = @gateway.send(:message_from, @gateway.send(:parse, successful_purchase_response))
+    assert_equal message, 'Succeeded'
+  end
+
+  def test_message_from_failed_transaction_response
+    message = @gateway.send(:message_from, declined_errors)
+    assert_equal message, 'gateway_error_message: DECLINED | gateway_response_errors: [gateway - DECLINED]'
+  end
+
+  def test_message_from_failed_transaction_response_412
+    message = @gateway.send(:message_from, { 'success' => false, 'errors' => ['fund_address_unspecified'], 'response_code' => 999 })
+    assert_equal message, 'errors: fund_address_unspecified'
+  end
+
+  def test_successful_authorize
+    billing_address = @options[:billing_address]
+    @gateway.expects(:ssl_request).with(
+      :post,
+      'https://uat.versapay.com/api/gateway/v1/orders/auth',
+      all_of(
+        includes(
+          "amount_cents\":\"#{@amount}",
```

Comment:
💬 Comment:

I don't know, I have mixed thoughts here, in one hand, I appreciate the DRY on the approach and how clean it looks, in the other side if one of these 14 params fails, the tests are not going tell you specifically which one is failing is just going to raise an "unexpected invocation"

Reply:
Right, I'm going to use the current check_request method.